src/ngx_http_lua_kong_module.c

```diff
@@ -670,5 +683,223 @@ ngx_http_lua_kong_get_upstream_ssl_verify(ngx_http_request_t *r,
+{
+    ngx_uint_t                  i;
+    ngx_http_variable_t        *v;
+    ngx_str_t                   name;
+    ngx_http_core_main_conf_t  *cmcf;
+
+    name.data = name_data;
+    name.len = name_len;
+
+    cmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_core_module);
+
+    /* from src/http/ngx_http_variables.c:ngx_http_get_variable_index */
+    v = cmcf->variables.elts;
+
+    if (v == NULL) {
+        /*
+         * this function will always be called post config,
+         * so we can assume the variable array always exists
+         */
+        *err = "can't access variables array";
+        return NGX_ERROR;
+
+    } else {
+        for (i = 0; i < cmcf->variables.nelts; i++) {
+            if (name.len != v[i].name.len
+                || ngx_strncasecmp(name.data, v[i].name.data, name.len) != 0)
+            {
+                continue;
+            }
+
+            *index = i;
+            return NGX_OK;
+        }
+    }
+
+    v = ngx_array_push(&cmcf->variables);
+    if (v == NULL) {
+        *err = "no memory";
+        return NGX_ERROR;
+    }
+
+    v->name.len = name.len;
+    v->name.data = ngx_pnalloc(ngx_cycle->pool, name.len);
+    if (v->name.data == NULL) {
+        *err = "no memory";
+        return NGX_ERROR;
+    }
+
+    ngx_strlow(v->name.data, name.data, name.len);
+
+    v->set_handler = NULL;
+    v->get_handler = NULL;
+    v->data = 0;
+    v->flags = 0;
+    v->index = cmcf->variables.nelts - 1;
+
+    *index = v->index;
+
+    return NGX_OK;
+}
+
+int
+ngx_http_lua_kong_ffi_var_get_by_index(ngx_http_request_t *r, ngx_uint_t index,
+    u_char **value, size_t *value_len, char **err)
+{
+    ngx_http_variable_value_t   *vv;
+
+    if (r == NULL) {
+        *err = "no request object found";
+        return NGX_ERROR;
+    }
+
+    if ((r)->connection->fd == (ngx_socket_t) -1) {
+        *err = "API disabled in the current context";
+        return NGX_ERROR;
+    }
+
+    vv = ngx_http_get_indexed_variable(r, index);
+    if (vv == NULL || vv->not_found) {
+        return NGX_DECLINED;
+    }
+
+    *value = vv->data;
+    *value_len = vv->len;
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_kong_ffi_var_set_by_index(ngx_http_request_t *r, ngx_uint_t index,
+    u_char *value, size_t value_len, char **err)
+{
+    u_char                      *p;
+    ngx_http_variable_t         *v;
+    ngx_http_variable_value_t   *vv;
+    ngx_http_core_main_conf_t   *cmcf;
 
-#endif /* NGX_HTTP_SSL */
+    if (r == NULL) {
+        *err = "no request object found";
+        return NGX_ERROR;
+    }
+
+    if ((r)->connection->fd == (ngx_socket_t) -1) {
+        *err = "API disabled in the current context";
+        return NGX_ERROR;
+    }
+
+    cmcf = ngx_http_get_module_main_conf(r, ngx_http_core_module);
+
+    if (index >= cmcf->variables.nelts) {
+       *err = "variable index out of range";
+        return NGX_ERROR;
+    }
+
+    v = ((ngx_http_variable_t*) cmcf->variables.elts) + index;
```

Comment:
`ngx_http_get_indexed_variable` returns the `ngx_http_variable_value_t` while setting the variable requires
the `ngx_http_variable_t` since the set handler needs to be called. [ngx_http_get_indexed_variable](https://github.com/nginx/nginx/blob/master/src/http/ngx_http_variables.c#L605) actually also first
get the `ngx_http_variable_t` then invokes `get_handler` to return the value similar to the flow we did here.